### ISSUE TITLE
Potential regression in 1.1.222: error: Expression of type "date* | int* | date" cannot be assigned to return type "DateT@get_today" (reportGeneralTypeIssues)

### ISSUE THREAD

**Issue by etfdev** [microsoft/pyright#3055]
I have inherited this piece of legacy code:

```python
import datetime as dt
from typing import Type, TypeVar

DateT = TypeVar("DateT", dt.date, int)


def _format_date_with_desired_type(date: dt.date, desired_type: Type[DateT]) -> DateT:
    if issubclass(desired_type, dt.date):
        return date  # type: ignore
    if issubclass(desired_type, int):  # type: ignore
        return int(date.strftime("%Y%m%d"))  # type: ignore
    raise TypeError(f"Cannot convert {date} to type {desired_type}")


def get_today(dtype: Type[DateT] = dt.date) -> DateT:
    today = dt.datetime.today()
    return _format_date_with_desired_type(today, desired_type=dtype)
```

This code produces no errors (which have not been ignored) in pyright 1.1.221, but I get the following in pyright 1.1.222:

```
$ PYRIGHT_PYTHON_FORCE_VERSION=1.1.222 pyright poc.py
/home/etfdev/poc.py
  /home/etfdev/poc.py:17:12 - error: Expression of type "date* | int* | date" cannot be assigned to return type "DateT@get_today"
    Type "date* | int* | date" cannot be assigned to type "DateT@get_today" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Is this a regression or the correct behavior in this case?

**Comment by erictraut** [comment]
This is intended behavior, although I agree it's not ideal. If anyone has a better idea, I'm open to it.

The previous versions of pyright contained a bug in how they dealt with a parameter annotated with a TypeVar type and a default argument, like `dtype` in your example. Mypy doesn't allow this combination because of the type challenges it poses. Pyright has allowed it for a while, but it contained a type analysis hole that was pointed out in [this issue](https://github.com/microsoft/pyright/issues/3036). My fix for this was to treat the input parameter as union of the TypeVar type and the type of the default argument. I believe this fixes the type analysis hole, but it's also awkward. Maybe it would be better to simply disallow this construct like mypy does?

**Comment by etfdev** [comment]
I have refactored the code so that `_format_date_with_desired_type()` is not used at all and am no longer affected by this change in behavior. Feel free to close this issue.
